### ISSUE TITLE
Enable to use token-authentication for Google Cloud Storage integration

### ISSUE THREAD

**Issue by komatsuna4747** [ropensci/targets#866]
## Prework

* [x] Read and agree to the [code of conduct](https://ropensci.org/code-of-conduct/) and [contributing guidelines](https://github.com/ropensci/targets/blob/main/CONTRIBUTING.md).
* [x] If there is [already a relevant issue](https://github.com/ropensci/targets/issues), whether open or closed, comment on the existing thread instead of posting a new issue.
* [x] New features take time and effort to create, and they take even more effort to maintain. So if the purpose of the feature is to resolve a struggle you are encountering personally, please consider first posting a "trouble" or "other" issue so we can discuss your use case and search for existing solutions first.
* [x] Format your code according to the [tidyverse style guide](https://style.tidyverse.org/).

## Proposal

Currently, when we use Google Cloud Storage for `{targets}`, it only seems to support authentication by a service account json file taken from a Google project.

https://github.com/ropensci/targets/blob/67c7444efab979618f6b0a6dc0d6b9ab8221e196/R/utils_gcp.R#L137-L141

It would be nice if we could use token-authentication using `gargle::token_fetch()` as shown below.

```r
googleCloudStorageR::gcs_auth(
  token = gargle::token_fetch(
    scopes = c("https://www.googleapis.com/auth/cloud-platform"),
    path = Sys.getenv("GCS_AUTH_FILE")
  )
)
```

Reference: https://code.markedmondson.me/googleCloudStorageR/articles/googleCloudStorageR.html

An advantage of this approach is that it encompasses both service account and user account authentication.

If you would like to authenticate as a service account using a key json file, you can do it by passing the path to the json file to `gargle:token_fetch()`:
```r
gargle::token_fetch(
  scopes = c("https://www.googleapis.com/auth/cloud-platform"),
  path = "/path/to/your/service-account.json"
)
```
Instead, if you would like to authenticate as a user account (this is what I use), you first run the following: 

```bash
$ gcloud auth application-default login
```
You will then be redirected to a web browser, where you can authenticate to GCP with your personal credentials, and the token will be stored in your local (`~/.config/gcloud/application_default_credentials.json` on not-Windows).
`gargle::token_fetch()` fetches info for authentication from the json file.

How gargle gets tokens: https://gargle.r-lib.org/articles/how-gargle-gets-tokens.html

**Comment by wlandau** [comment]
Thanks for suggesting. Supplying a token through `token_fetch()` looks like it would add convenience and act in line with the intent usage of `googleCloudStorageR`. @MarkEdmondson1234, are there any usability or security concerns we are missing here?

> You will then be redirected to a web browser, where you can authenticate to GCP with your personal credentials, and the token will be stored in your local (~/.config/gcloud/application_default_credentials.json on not-Windows).
gargle::token_fetch() fetches info for authentication from the json file.

`targets` works best when a target does not require interactive real-time input from the user (i.e. allows the pipeline to be fully automated). Does this mean this part should be documented in `targets` somewhere so users know to download the token before running the pipeline?

**Comment by MarkEdmondson1234** [comment]
User based tokens is probably the most common way to authenticate since some people only get user level access to the buckets if they didn't create them. It's usually interactive only in the initial flow (best out of targets) then gargle takes care of refresh. 

In a production setting I would still recommend the service key approach but there is no denying the convenience of user based tokens. 

The absolute best if your R code is executing on GCP is to use application default credentials that are embedded within the service so you don't need to worry about auth files at all. This is supported via gargle. 

I would recommend leaning on gargle to handle authentication as much as possible and to not roll your own, as there are pretty regular updates that it handles for the majority of GCP R packages.

**Comment by wlandau** [comment]
A PR would be great. But first, given Mark's comment, I wonder if we should let the user choose the authentication method, with service-account-based auth as the default. Something like:

```r
gcp_gcs_auth <- function(verbose = FALSE, authentication = "service") { 
  if_any(verbose, identity, suppressMessages) (
    switch(
      authentication,
      service = googleCloudStorageR::gcs_auth(Sys.getenv("GCS_AUTH_FILE")),
      token = googleCloudStorageR::gcs_auth(
        token = gargle::token_fetch(
          scopes = c("https://www.googleapis.com/auth/cloud-platform"),
          path = Sys.getenv("GCS_AUTH_FILE")
        )
      )
    )
  ) 
} 
```

Depending on what we decide for `gcp_gcs_auth()`, I will then modify the interface to allow the choice of auth method via `tar_option_set()`:

```
# _targets.R
# ...
tar_option_set(
  resources = tar_resources(
    gcp = tar_resources_gcp(authentication = "token")
  )
)
# ...
```

After that, would you be willing to submit a PR to https://github.com/ropensci-books/targets/blob/e798e3ce820c8717d08418078299b03690431b22/data.Rmd#L109-L120 describing how to choose a suitable authentication method (if applicable) and how to do the up-front interactive part before running the pipeline?

**Comment by komatsuna4747** [comment]
I'd be happy to make pull requests for this implementation and documentation :)

Regarding `gcp_gcs_auth()`, I was wondering if there was a benefit to make `service` and `token` selectable.
According to [this vignette](https://gargle.r-lib.org/articles/how-gargle-gets-tokens.html), `gargle::token_fetch()` first looks for service account credentials. If it fails, it then starts searching for other credentials, including user based ones.

I thought using `token` would be enough to cover both service account and user account authentication:

```r
gcp_gcs_auth <- function(verbose = FALSE) { 
  if_any(verbose, identity, suppressMessages) (
    googleCloudStorageR::gcs_auth(
      token = gargle::token_fetch(
        scopes = c("https://www.googleapis.com/auth/cloud-platform"),
        path = Sys.getenv("GCS_AUTH_FILE")
      )
    )
  ) 
} 
```

If one is sure to use service account authentication, it might be safer to keep `service`.

@MarkEdmondson1234 
I'd appreciate it if you could give us your opinion.

**Comment by MarkEdmondson1234** [comment]
Yep token_fetch() will do as you say, I agree with you it would be best to leave the logic within that function as much as possible. A choice between service account and token makes sense as I can imagine during development it's a user token, then switches to a service account (if not on GCP) during production, but other auth decisions is best to leave within gargle config.

**Comment by wlandau** [comment]
Thanks for clarifying, @MarkEdmondson1234. @komatsuna4747, please feel free to proceed with a PR based on your original suggestion in https://github.com/ropensci/targets/issues/866#issue-1298981463.

**Comment by wlandau** [comment]
Fixed in #868

**Comment by komatsuna4747** [comment]
I forgot to update the manual for the authentication ways. I will soon submit a pull request for that.
